#!/bin/bash
#PBS -N ZB_read_depth
#PBS -j oe
#PBS -m ae
#PBS -l nodes=1
#PBS -l ncpus=21
#PBS -l mem=100gb
#PBS -l walltime=24:00:00
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Define paths to samtools"
export PATH=$PATH:/usr/local/miniconda3/envs/assembly/bin/

echo "---------------------------------------"
echo 'DIRECT ALIGNMENT'

echo "---------------------------------------"
echo 'Change to working directory and set env variables'
cd /home/jwestaway/pk_pipeline/ZB_100/outputs/direct_alignment/
BED='/home/jwestaway/pk_pipeline/ref_genomes/PKA1H1/strain_A1_H.1.Icor.fasta.bed'

echo "---------------------------------------"
echo 'Sort bam files'

for i in *.bam
do
samtools sort -@ 20 $i > ${i%.bam}.sorted.bam
done 

echo "---------------------------------------"
echo 'Index bam files'

for i in *.sorted.bam
do
samtools index -@ 20 $i 
done 

echo "---------------------------------------"
echo 'Create file with list of filenames'
ls *sorted.bam > filenames.txt

echo "---------------------------------------"
echo 'Exectue samtools depth'
samtools depth -a -b $BED -f filenames.txt -o ZB_redo_direct_depth_summary.depth

echo "---------------------------------------"
echo 'INDIRECT ALIGNMENT'

echo "---------------------------------------"
echo 'Change to working directory and set env variables'
cd /home/jwestaway/pk_pipeline/ZB_100/outputs/hg_removed_alignment/pk_alignment/

echo "---------------------------------------"
echo 'Sort bam files'

for i in *.bam
do
samtools sort -@ 20 $i > ${i%.bam}.sorted.bam
done 

echo "---------------------------------------"
echo 'Index bam files'

for i in *.sorted.bam
do
samtools index -@ 20 $i 
done 

echo "---------------------------------------"
echo 'Create file with list of filenames'
ls *sorted.bam > filenames.txt

echo "---------------------------------------"
echo 'Exectue samtools depth'
samtools depth -a -b $BED -f filenames.txt -o ZB_redo_indirect_depth_summary.depth

echo "---------------------------------------"
echo "Finsihed!"