#!/bin/bash
#PBS -N Mixed_ploidy_test
#PBS -j oe
#PBS -m ae
#PBS -l nodes=1
#PBS -l ncpus=21
#PBS -l mem=100gb
#PBS -l walltime=48:00:00
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Define paths to samtools"
export PATH=$PATH:/usr/local/jdk1.8.0_131/bin
export PATH=$PATH:/usr/local/jre1.8.0_111/bin
export PATH=$PATH:/usr/local/miniconda3/envs/assembly/bin/

echo "---------------------------------------"
echo 'Set env variables'
OUTDIR="/home/jwestaway/pk_pipeline/ZB_100/outputs/vcf_consensus/GATK/ploidy_test"
PICARD="/usr/local/miniconda3/pkgs/picard-2.21.9-0/share/picard-2.21.9-0/picard.jar"
INDEXTDIR="/home/jwestaway/pk_pipeline/ref_genomes/PKA1H1/fasta/strain_A1_H.1.Icor.fasta"
INTERVALS="/home/jwestaway/pk_pipeline/ref_genomes/PKA1H1/strain_A1_H.1.Icor.fasta.bed"

mkdir $OUTDIR
cd /home/jwestaway/pk_pipeline/ZB_100/outputs/vcf_consensus/GATK


echo "---------------------------------------"
echo 'SAMPLE 1'
echo "---------------------------------------"


echo "---------------------------------------"
echo 'HaplotypeCaller: haploid' 

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /home/jwestaway/pk_pipeline/tools/gatk-4.2.2.0/gatk-package-4.2.2.0-local.jar \
    HaplotypeCaller \
    -ploidy 1 \
    -ERC GVCF \
    --num-pruning-samples 3 \
    --max-num-haplotypes-in-population 200 \
    -G StandardAnnotation \
    -G AS_StandardAnnotation \
    -contamination 0.0 \
    -R $INDEXTDIR \
    --intervals $INTERVALS \
    -I PK_SB_DNA_006_DKDL210002135-1a_HWHGKDSXY_L4/PK_SB_DNA_006_DKDL210002135-1a_HWHGKDSXY_L4.dupmarked.realigned.recal.bam \
    -O $OUTDIR/PK_SB_DNA_006_DKDL210002135-1a_HWHGKDSXY_L4.haploid.vcf.gz

echo "---------------------------------------"
echo 'HaplotypeCaller: diploid' 

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /home/jwestaway/pk_pipeline/tools/gatk-4.2.2.0/gatk-package-4.2.2.0-local.jar \
    HaplotypeCaller \
    -ERC GVCF \
    -ploidy 2 \
    --num-pruning-samples 3 \
    --max-num-haplotypes-in-population 200 \
    -contamination 0.0 \
    -G StandardAnnotation \
    -G AS_StandardAnnotation \
    -R $INDEXTDIR \
    --intervals $INTERVALS \
    --heterozygosity 0.0259 \
    --indel-heterozygosity 0.0032 \
    -I PK_SB_DNA_006_DKDL210002135-1a_HWHGKDSXY_L4/PK_SB_DNA_006_DKDL210002135-1a_HWHGKDSXY_L4.dupmarked.realigned.recal.bam \
    -O $OUTDIR/PK_SB_DNA_006_DKDL210002135-1a_HWHGKDSXY_L4.diploid.vcf.gz

echo "---------------------------------------"
echo 'MergeVCFs' 
/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -jar $PICARD \
    MergeVcfs \
    I=$OUTDIR/PK_SB_DNA_006_DKDL210002135-1a_HWHGKDSXY_L4.haploid.vcf.gz \
    I=$OUTDIR/PK_SB_DNA_006_DKDL210002135-1a_HWHGKDSXY_L4.diploid.vcf.gz \
    O=$OUTDIR/PK_SB_DNA_006_DKDL210002135-1a_HWHGKDSXY_L4_merged.vcf.gz

echo "---------------------------------------"
echo 'Finished' 