#!/bin/bash
#PBS -N Variant_calling_gatk_test
#PBS -j oe
#PBS -m ae
#PBS -l nodes=1
#PBS -l ncpus=21
#PBS -l mem=150gb
#PBS -l walltime=96:00:00
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Define paths"
export PATH=$PATH:/usr/local/jdk1.8.0_131/bin
export PATH=$PATH:/usr/local/jre1.8.0_111/bin
export PATH=$PATH:/usr/local/GenomeAnalysisTK-3.2.2 
export PATH=$PATH:/usr/local/miniconda3/envs/assembly/bin/


echo "---------------------------------------"
echo "---------------------------------------"
echo 'Change to working directory and set env variables'
PICARD="/usr/local/miniconda3/pkgs/picard-2.21.9-0/share/picard-2.21.9-0/picard.jar"
INDEXTDIR="/home/jwestaway/pk_pipeline/ref_genomes/PKA1H1/fasta/strain_A1_H.1.Icor.fasta"
OUTDIR="/home/jwestaway/pk_pipeline/Initial_Pk_subset/outputs/variant_calling/GATK/TEST"
INTERVALS="/home/jwestaway/pk_pipeline/ref_genomes/PKA1H1/strain_A1_H.1.Icor.fasta.bed"

cd $OUTDIR

echo "---------------------------------------"
echo 'HaplotypeCaller' 

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /usr/local/GenomeAnalysisTK-3.2.2/GenomeAnalysisTK.jar \
    -T HaplotypeCaller \
    -ERC GVCF \
    --minPruning 3 \
    --maxNumHaplotypesInPopulation 200 \
    --max_alternate_alleles 3 \
    --variant_index_type LINEAR \
    --variant_index_parameter 128000 \
    -contamination 0.0 \
    -G Standard \
    -R $INDEXTDIR \
    -I PK_SB_DNA_001_DKDL210002130-1a_HWHGKDSXY_L4.dupmarked.realigned.recal.bam \
    -o PK_SB_DNA_001_DKDL210002130-1a_HWHGKDSXY_L4.haplotypecaller.g.vcf.gz

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /usr/local/GenomeAnalysisTK-3.2.2/GenomeAnalysisTK.jar \
    -T HaplotypeCaller \
    -ERC GVCF \
    --minPruning 3 \
    --maxNumHaplotypesInPopulation 200 \
    --max_alternate_alleles 3 \
    --variant_index_type LINEAR \
    --variant_index_parameter 128000 \
    -contamination 0.0 \
    -G Standard \
    -R $INDEXTDIR \
    -I PK_SB_DNA_002_DKDL210002131-1a_HWHGKDSXY_L4.dupmarked.realigned.recal.bam \
    -o PK_SB_DNA_002_DKDL210002131-1a_HWHGKDSXY_L4.haplotypecaller.g.vcf.gz


echo "---------------------------------------"
echo 'CombineGVCFs'
/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /usr/local/GenomeAnalysisTK-3.2.2/GenomeAnalysisTK.jar \
    -T CombineGVCFs \
    -R $INDEXTDIR \
    -V PK_SB_DNA_001_DKDL210002130-1a_HWHGKDSXY_L4.haplotypecaller.g.vcf.gz -V PK_SB_DNA_002_DKDL210002131-1a_HWHGKDSXY_L4.haplotypecaller.g.vcf.gz \
    -o Combined.g.vcf.gz

echo "---------------------------------------"
echo 'GenotypeGVCFs' 
/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /usr/local/GenomeAnalysisTK-3.2.2/GenomeAnalysisTK.jar \
    -T GenotypeGVCFs \
    -nt 20 \
    -R $INDEXTDIR \
    -V Combined.g.vcf.gz \
    -o Genotyped.vcf.gz

echo "---------------------------------------"
echo "Finished "
