#!/bin/bash
#PBS -N Annotate_variants
#PBS -j oe
#PBS -m ae
#PBS -l nodes=1
#PBS -l ncpus=16
#PBS -l mem=50gb
#PBS -l walltime=24:00:00
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "---------------------------------------"
echo 'Change to working directory and set env variables'
export PATH=$PATH:/usr/local/jdk1.8.0_131/bin
export PATH=$PATH:/usr/local/jre1.8.0_111/bin
OUTDIR="/home/jwestaway/pk_pipeline/ZB_100/outputs/vcf_consensus/variant_annotation"
GFF="/home/jwestaway/pk_pipeline/ref_genomes/PKA1H1/gff/"
FASTA="/home/jwestaway/pk_pipeline/ref_genomes/PKA1H1/fasta/strain_A1_H.1.Icor.fasta"
SNPEF="/home/jwestaway/pk_pipeline/tools/snpEff"


echo "---------------------------------------------------------------------------------------------------------------------"
echo "Build DB for Pk"
cd $SNPEF

mv $GFF/strain_A1_H.1.Icor.gff3 $SNPEF/data/PKA1H1/regulation.gff.gz


/usr/local/jdk1.8.0_131/bin/java -jar snpEff.jar build -gff3 -v PKA1H1



echo "---------------------------------------------------------------------------------------------------------------------"
echo "Annotate VCF"

echo "---------------------------------------"
echo "FINISHED"