#!/bin/bash
#PBS -N Plot_annotations
#PBS -j oe
#PBS -m ae
#PBS -l nodes=1
#PBS -l ncpus=16
#PBS -l mem=50gb
#PBS -l walltime=24:00:00
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "---------------------------------------"
echo 'Change to working directory and set env variables'
OUTDIR="/home/jwestaway/pk_pipeline/ZB_100/outputs/vcf_consensus/annotation_distribution"
PICARD="/usr/local/miniconda3/pkgs/picard-2.21.9-0/share/picard-2.21.9-0/picard.jar"
INDEXTDIR="/home/jwestaway/pk_pipeline/ref_genomes/PKA1H1/fasta/strain_A1_H.1.Icor.fasta"

mkdir $OUTDIR

cd /home/jwestaway/pk_pipeline/ZB_100/outputs/vcf_consensus/

echo "---------------------------------------"
echo 'Make diagnostic tables for Variants Scores - snps'

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /home/jwestaway/pk_pipeline/tools/gatk-4.2.2.0/gatk-package-4.2.2.0-local.jar \
    VariantsToTable \
    -R $INDEXTDIR \
    -V GVCFall_SNPs.vcf \
    -F CHROM -F POS -F QUAL -F QD -F DP -F MQ -F GQ -F MQRankSum -F FS -F ReadPosRankSum -F SOR \
    -O $OUTDIR/GVCFall_SNPs.table


echo "---------------------------------------"
echo 'Make diagnostic tables for Variants Scores - indels'

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /home/jwestaway/pk_pipeline/tools/gatk-4.2.2.0/gatk-package-4.2.2.0-local.jar \
    VariantsToTable \
    -R $INDEXTDIR \
    -V GVCFall_INDELs.vcf \
    -F CHROM -F POS -F QUAL -F QD -F DP -F MQ -F GQ -F MQRankSum -F FS -F ReadPosRankSum -F SOR \
    -O $OUTDIR/GVCFall_INDELs.table


echo "---------------------------------------"
echo 'Finsihed'