#!/bin/bash
#PBS -N Filter_variants
#PBS -j oe
#PBS -m ae
#PBS -l nodes=1
#PBS -l ncpus=16
#PBS -l mem=50gb
#PBS -l walltime=12:00:00
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Define paths and change to working directory"
INDEXTDIR="/home/jwestaway/pk_pipeline/ref_genomes/PKA1H1/fasta/strain_A1_H.1.Icor.fasta"
INDIR="/home/jwestaway/pk_pipeline/Initial_Pk_subset/outputs/variant_calling/consensus"
OUTDIR="/home/jwestaway/pk_pipeline/Initial_Pk_subset/outputs/variant_calling/filtered"
PICARD="/usr/local/miniconda3/pkgs/picard-2.21.9-0/share/picard-2.21.9-0/picard.jar"
INDEXTDIR="/home/jwestaway/pk_pipeline/ref_genomes/PKA1H1/fasta/strain_A1_H.1.Icor.fasta"

cd $INDIR

echo "---------------------------------------"
echo 'SELECT VARIANT TYPES'
echo "---------------------------------------"

echo "---------------------------------------"
echo "SNPs"

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /home/jwestaway/pk_pipeline/tools/gatk-4.2.2.0/gatk-package-4.2.2.0-local.jar \
    SelectVariants \
    -R $INDEXTDIR \
    -V Initial_Pk_dataset_consensus.vcf \
    --select-type-to-include SNP \
    -O $OUTDIR/GVCFall_SNPs.vcf

echo "---------------------------------------"
echo "Indels"

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /home/jwestaway/pk_pipeline/tools/gatk-4.2.2.0/gatk-package-4.2.2.0-local.jar \
    SelectVariants \
    -R $INDEXTDIR \
    -V Initial_Pk_dataset_consensus.vcf \
    --select-type-to-include INDEL \
    -O $OUTDIR/GVCFall_INDELs.vcf

echo "---------------------------------------"
echo 'FILTER'
echo "---------------------------------------"

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /home/jwestaway/pk_pipeline/tools/gatk-4.2.2.0/gatk-package-4.2.2.0-local.jar \
    VariantFiltration \
    -V $INDIR/Initial_Pk_dataset_consensus.vcf \
    -filter "QD < 20.0" --filter-name "QD2" \
    -filter "FS > 2.0" --filter-name "FS60" \
    -filter "MQ < 59.0" --filter-name "MQ40" \
    -O $OUTDIR/PK_consensus_FILTERED.vcf

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /home/jwestaway/pk_pipeline/tools/gatk-4.2.2.0/gatk-package-4.2.2.0-local.jar \
    VariantFiltration \
    -V $OUTDIR/GVCFall_SNPs.vcf \
    -filter "QD < 20.0" --filter-name "QD2" \
    -filter "FS > 2.0" --filter-name "FS60" \
    -filter "MQ < 59.0" --filter-name "MQ40" \
    -O $OUTDIR/FILTERED_SNPs.vcf

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /home/jwestaway/pk_pipeline/tools/gatk-4.2.2.0/gatk-package-4.2.2.0-local.jar \
    VariantFiltration \
    -V $OUTDIR/GVCFall_INDELs.vcf \
    -filter "QD < 20.0" --filter-name "QD2" \
    -filter "FS > 2.0" --filter-name "FS60" \
    -filter "MQ < 59.0" --filter-name "MQ40" \
    -O $OUTDIR/FILTERED_INDELs.vcf

echo "---------------------------------------"
echo "Finsihed"
