#!/bin/bash
#PBS -N merge_vcfs
#PBS -j oe
#PBS -m ae
#PBS -l nodes=1
#PBS -l ncpus=16
#PBS -l mem=100gb
#PBS -l walltime=48:00:00
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Define paths to samtools"
export PATH=$PATH:/usr/local/jdk1.8.0_131/bin
export PATH=$PATH:/usr/local/jre1.8.0_111/bin
export PATH=$PATH:/usr/local/GenomeAnalysisTK-3.2.2 

echo "---------------------------------------"
echo 'Change to working directory and set env variables'
cd /home/jwestaway/pk_pipeline/ZB_100/outputs/vcf_consensus/GATK/
PICARD="/usr/local/miniconda3/pkgs/picard-2.21.9-0/share/picard-2.21.9-0/picard.jar"

echo "---------------------------------------"
echo 'Create list of file names from GATK variant calling'
ls /home/jwestaway/pk_pipeline/ZB_100/outputs/vcf_consensus/GATK/*/*.vcf.gz \
    | tr '\n' '\0' | xargs -0 -n 1 basename \
    > input_variant_files.list

echo "---------------------------------------"
echo 'Execute MergeVcfs'
/usr/local/jdk1.8.0_131/bin/java -jar $PICARD MergeVcfs \
    I=input_variant_files.list \
    O=GATK_merged_variants.vcf.gz

echo "---------------------------------------"
echo 'Finished' 