#!/bin/bash
#PBS -N Bam_pre-processing
#PBS -j oe
#PBS -m ae
#PBS -l nodes=1
#PBS -l ncpus=11
#PBS -l mem=50gb
#PBS -l walltime=12:00:00
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Define paths to samtools"
export PATH=$PATH:/usr/local/jdk1.8.0_131/bin
export PATH=$PATH:/usr/local/jre1.8.0_111/bin
export PATH=$PATH:/usr/local/miniconda3/envs/assembly/bin/

echo "---------------------------------------"
echo "---------------------------------------"
echo 'Change to working directory and set env variables'

INDIR="/home/jwestaway/pk_pipeline/Initial_Pk_subset/outputs/batch5/pk_alignment"
OUTDIR="/home/jwestaway/pk_pipeline/Initial_Pk_subset/outputs/variant_calling/batch5"
PICARD="/usr/local/miniconda3/pkgs/picard-2.21.9-0/share/picard-2.21.9-0/picard.jar"
INDEXTDIR="/home/jwestaway/pk_pipeline/ref_genomes/PKA1H1/fasta/strain_A1_H.1.Icor.fasta"
INTERVALS="/home/jwestaway/pk_pipeline/ref_genomes/PKA1H1/strain_A1_H.1.Icor.fasta.bed"
KNOWNSITES="/home/jwestaway/pk_pipeline/ref_genomes/variants"
 
mkdir $OUTDIR
mkdir $OUTDIR/final_bam

echo "---------------------------------------"
echo 'Sort and Index bam files'

cd /home/jwestaway/pk_pipeline/Initial_Pk_subset/outputs/batch5/pk_alignment

echo "---------------------------------------"
echo 'MarkDuplicates'

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -jar $PICARD \
    MarkDuplicates AS=TRUE VALIDATION_STRINGENCY=LENIENT \
    I=$INDIR/SAMPLE.sorted.bam \
    O=$OUTDIR/SAMPLE.dupmarked.bam \
    M=$OUTDIR/SAMPLE_picard_metrics_file.txt 

echo "---------------------------------------"
echo 'Change header @RG and index' 

samtools view -H $OUTDIR/SAMPLE.dupmarked.bam | \
    sed 's,^@RG.*,@RG\tID:SAMPLE\tSM:SAMPLE\tLB:None\tPL:Illumina,g' |  \
    samtools reheader - $OUTDIR/SAMPLE.dupmarked.bam > $OUTDIR/SAMPLE.dupmarked.reheader.bam

samtools index $OUTDIR/SAMPLE.dupmarked.reheader.bam 

echo "---------------------------------------"
echo 'RealignerTargetCreator'

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /usr/local/GenomeAnalysisTK-3.2.2/GenomeAnalysisTK.jar \
    -T RealignerTargetCreator \
    -nt 10 \
    -R $INDEXTDIR \
    -I $OUTDIR/SAMPLE.dupmarked.reheader.bam \
    --intervals $INTERVALS \
    -known $KNOWNSITES/TRUE_INDELs.vcf \
    -o $OUTDIR/SAMPLE.dupmarked.realigner.intervals

echo "---------------------------------------"
echo 'IndelRealigner' 

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /usr/local/GenomeAnalysisTK-3.2.2/GenomeAnalysisTK.jar \
    -T IndelRealigner \
    --consensusDeterminationModel KNOWNS_ONLY \
    -LOD 0.4 \
    -R $INDEXTDIR \
    -I $OUTDIR/SAMPLE.dupmarked.reheader.bam \
    --intervals $INTERVALS \
    -known $KNOWNSITES/TRUE_INDELs.vcf \
    -targetIntervals $OUTDIR/SAMPLE.dupmarked.realigner.intervals \
    -o $OUTDIR/SAMPLE.dupmarked.realigned.bam

echo "---------------------------------------"
echo 'BaseRecalibrator - create recal table' 

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /usr/local/GenomeAnalysisTK-3.2.2/GenomeAnalysisTK.jar \
    -T BaseRecalibrator \
    -R $INDEXTDIR \
    -I $OUTDIR/SAMPLE.dupmarked.realigned.bam \
    --intervals $INTERVALS \
    -knownSites $KNOWNSITES/TRUE_consensus.vcf \
    -o $OUTDIR/SAMPLE.dupmarked.realigned.recal.table

echo "---------------------------------------"
echo 'PrintReads - get recal reads' 

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /usr/local/GenomeAnalysisTK-3.2.2/GenomeAnalysisTK.jar  \
    -T PrintReads \
    -R $INDEXTDIR \
    --intervals $INTERVALS \
    -I $OUTDIR/SAMPLE.dupmarked.realigned.bam \
    -BQSR $OUTDIR/SAMPLE.dupmarked.realigned.recal.table \
    -o $OUTDIR/final_bam/SAMPLE.dupmarked.realigned.recal.bam

echo "---------------------------------------"
echo 'Clean up' 

rm -f $OUTDIR/SAMPLE.dupmarked.bam
rm -f $OUTDIR/SAMPLE.dupmarked.realigned.recal.table
rm -f $OUTDIR/SAMPLE.dupmarked.realigned.bam
rm -f $OUTDIR/SAMPLE.dupmarked.realigner.intervals
rm -f $OUTDIR/SAMPLE.dupmarked.reheader.bam 
rm -f $OUTDIR/SAMPLE_picard_metrics_file.txt 

echo "---------------------------------------"
echo 'Finished' 


