#!/bin/bash
#PBS -N SNPs_&_indels_from_consesnus_tools
#PBS -j oe
#PBS -m ae
#PBS -l nodes=1
#PBS -l ncpus=21
#PBS -l mem=100gb
#PBS -l walltime=24:00:00
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Define paths and change to working directory"
INDEXTDIR="/home/jwestaway/pk_pipeline/ref_genomes/PKA1H1/fasta/strain_A1_H.1.Icor.fasta"
OUTDIR="/home/jwestaway/pk_pipeline/ZB_100/outputs/vcf_consensus/filtering_tests/tool_spec_consensus/"
GATK="/home/jwestaway/pk_pipeline/ZB_100/outputs/vcf_consensus/GATK/"
bcftools="/home/jwestaway/pk_pipeline/ZB_100/outputs/vcf_consensus/samtools/"


cd /home/jwestaway/pk_pipeline/ZB_100/outputs/vcf_consensus

echo "---------------------------------------------------------------------------------------------------------------------"
echo "SNPs and Indels from tool-specific conesnsus VCF files"
echo "---------------------------------------------------------------------------------------------------------------------"


echo "---------------------------------------"
echo "GATK"


echo "---------------------------------------"
echo "SNPs"

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /home/jwestaway/pk_pipeline/tools/gatk-4.2.2.0/gatk-package-4.2.2.0-local.jar \
    SelectVariants \
    -R $INDEXTDIR \
    -V $GATK/GATK_consensus.vcf \
    --select-type-to-include SNP \
    -O $OUTDIR/GVCFall_SNPs_GATK.vcf

echo "---------------------------------------"
echo "Indels"

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /home/jwestaway/pk_pipeline/tools/gatk-4.2.2.0/gatk-package-4.2.2.0-local.jar \
    SelectVariants \
    -R $INDEXTDIR \
    -V $GATK/GATK_consensus.vcf \
    --select-type-to-include INDEL \
    -O $OUTDIR/GVCFall_INDELs_GATK.vcf


echo "---------------------------------------"
echo "samtools/bcftools"


echo "---------------------------------------"
echo "SNPs"

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /home/jwestaway/pk_pipeline/tools/gatk-4.2.2.0/gatk-package-4.2.2.0-local.jar \
    SelectVariants \
    -R $INDEXTDIR \
    -V $bcftools/bcftools_consensus.vcf \
    --select-type-to-include SNP \
    -O $OUTDIR/GVCFall_SNPs_bcftools.vcf

echo "---------------------------------------"
echo "Indels"

/usr/local/jdk1.8.0_131/bin/java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar /home/jwestaway/pk_pipeline/tools/gatk-4.2.2.0/gatk-package-4.2.2.0-local.jar \
    SelectVariants \
    -R $INDEXTDIR \
    -V $bcftools/bcftools_consensus.vcf \
    --select-type-to-include INDEL \
    -O $OUTDIR/GVCFall_INDELs_bcftools.vcf

echo "---------------------------------------"
echo "Finished"